Adding development definition

bastilian opened the pull request:
This adds a separate compose definition for a development environment to start, which only consists of `foreman`, `proxy`, `db` and `client`. 

It also cleans up the bootstrapping process with switches to install dependencies based on an environment variable set in each definition as build arguments and environment variables to run the containers.

The `foreman_version` argument allow now to pass in a specific git branch to be cloned. Cloning now has set `--depth 1` to not clone history and keep the containers small.

Migrating and and bundling happens now on first start of a container and leaves a mark to skip it on following starts in development this is skipped all together to allow installing and bootstrapping manually.

Asset volume paths are now combined into `.asset-cache` and ignored together with pgdata directories.

To easily build and up the containers there is now a `Makefile` with tasks, which are listed in the `README.md`.

It still needs a bit of tweaking there are some quirks with bootstrapping the database.

shlomizadok commented:
@bastilian - thanks for this amazing improvement.
I have tried running both `make up` and `make up-development`. Both fail with "No such file or directory"

Any clue what am I doing wrong?
Thanks a lot!

bastilian commented:
@shlomizadok I tried reproducing this, even with empty docker registry and freshly cloned it still ups as it should. Mind pasting the whole log output?

ohadlevy commented on docker-compose.yml:
```diff
@@ -1,25 +1,35 @@
+    build:
+      context: foreman
+      args:
+        foreman_version: 1.13.1
```
this should be a branch name in github, or we need to change the git clone command to use a tag.

this caused my foreman image to fail, and a very interesting side note, when git fails it actually removed the /usr/src/foreman, which was ***very*** unexpected.

bastilian replied:
I'm gonna change this and use '1.13-stable' and split cloning and switching the branch to see if this prevents a the failure and removal of the directory.

ohadlevy commented on foreman/Dockerfile:
```diff
@@ -1,20 +1,30 @@
+MAINTAINER Ohad Levy "ohadlevy@gmail.com"
+
+ARG foreman_version
+ARG foreman_mode
```
why do we need foreman_mode and FOREMAN_MODE?

bastilian replied:
I've introduced this to have a common variable that the containers inherit from and use instead of relying on `RAILS_ENV` or similar.

ohadlevy commented on foreman/support/install-packages.sh:
```diff
@@ -0,0 +1,18 @@
+
+dnf clean all
+
+gem install bundler foreman puma
```
IMHO foreman gem is only needed for dev env?

bastilian replied:
Yes, I will move `foreman`.

ohadlevy commented on foreman/support/install.sh:
```diff
@@ -0,0 +1,22 @@
+
+  echo "=> Seeding database"
+  bundle exec rake db:seed
+
```
nitpick: you can run all rake commands in one go, this will reduce boot times: I do something like:
```sh
git pull && bundle update && npm update --no-optional && npm prune && \
 FOREMAN_APIPIE_LANGS=en RAILS_ENV=production bundle exec rake db:migrate db:seed \
 assets:clean assets:precompile webpack:compile apipie:cache:index &&  \
 touch tmp/restart.txt && `bundle show foreman-tasks`/bin/dynflow-executor restart
```

bastilian replied:
You are right! I'm gonna look more into combining needed rake tasks, so they run in one go.

shlomizadok commented:
@bastilian - The error I get is:
```bash
docker-compose -f docker-compose.yml up
foremandockercompose_db_1 is up-to-date
foremandockercompose_memcache_1 is up-to-date
foremandockercompose_proxy_1 is up-to-date
Creating foremandockercompose_foreman_1
ERROR: no such file or directory
Makefile:11: recipe for target 'up' failed
make: *** [up] Error 1
```

I have tried to debug which file is not found, without much success

bastilian commented:
@shlomizadok since @ohadlevy mentioned that his `/usr/src/foreman` was removed in the clone process, it could be that your setup has the same issue. But that is just a hunch for now.

I'm gonna rewrite the bootrapping and add verifications that everything that the container needs to boot is in place.

ripcurld00d commented:
@shlomizadok @ohadlevy @bastilian 
I tried too to deploy both the production and the development environments and didn't get any errors (I am using `docker-compose version 1.8.1, build 878cff1`).

However, `curl`-ing Foreman directly fails with the following message:
> We're sorry, but something went wrong.
> If you are the application owner check the logs for more information.

Reading the log file reveals that `rake webpack:compile` wasn't executed or failed:

>  | ActionView::Template::Error (Could not load compiled manifest from /usr/src/foreman/public/webpack/manifest.json - have you run `rake webpack:compile`? (original err
or No such file or directory @ rb_sysopen - /usr/src/foreman/public/webpack/manifest.json)):
 |     12: 
 |     13:     <%= favicon_link_tag "favicon.ico"%>
 |     14: 
 |     15:     <%= stylesheet_link_tag *webpack_asset_paths('bundle', :extension => 'css'), "data-turbolinks-track" => true %>
 |     16:     <%= stylesheet_link_tag 'application', "data-turbolinks-track" => true %>
 |     17:     <%= yield(:stylesheets) %>
 |     18: 
 |   app/views/layouts/base.html.erb:15:in `_app_views_layouts_base_html_erb___2388900686973559175_47033682369420'
 |   app/views/layouts/application.html.erb:5:in `_app_views_layouts_application_html_erb___3631922090979221865_47033692944200'
 |   app/controllers/application_controller.rb:317:in `generic_exception'
 |   lib/middleware/catch_json_parse_errors.rb:9:in `call'
 |   lib/middleware/tagged_logging.rb:18:in `call'

bastilian commented:
@ripcurld00d Not compiling assets in development is by design. I will add a section in the README elaborating on this.

shlomizadok commented:
![image](https://cloud.githubusercontent.com/assets/433583/20036328/81f115dc-a40e-11e6-898b-d4d1b666b244.png)
@github is so smart, it knows what you are going to say in the future !!

bastilian commented:
![screen shot 2016-11-06 at 09 49 52](https://cloud.githubusercontent.com/assets/7757/20036341/67c7bcae-a406-11e6-828e-4aa841203bf0.png)

@shlomizadok hahahah. that is gold! We are in the future! yes! we made it! :D

shlomizadok commented:
Okay, to make this work we need `docker --version == 1.12.3` and 
`docker-compose --version == 1.8.1`
@bastilian - maybe we should check on the Makefile for those versions?

shlomizadok commented on Makefile:
```diff
@@ -0,0 +1,14 @@
+up:
+	docker-compose -f docker-compose.yml up
+
+up-development:
```
`up_development` ? (e.g., with underscore to keep consistency)

shlomizadok commented on foreman/support/install.sh:
```diff
@@ -0,0 +1,22 @@
+  echo "=> Running install.sh"
+
+  echo "=> Installing npm packages"
+  npm install
```
Please add `npm install npm` as a step before - to update npm first, else it would fail on some packages

bastilian commented on Makefile:
```diff
@@ -0,0 +1,14 @@
+
+up:
+	docker-compose -f docker-compose.yml up
+
```
Oh, yes! a test container! on my list. :)

bastilian commented:
@shlomizadok & @ohadlevy Will work on the changes today and squash my commits for final review and testing.

bastilian commented:
I've ironed out the remaining issues.

The packages should now be at minimum to have a small container (they are still around 600 - 700 MB though). 

Gems and node_modules are installed only on build now, as does compiling assets.
Bootstrapping the database happens on first boot and removes the bootstrap mark.

The README describes the tasks available better now. And states the required docker and compose version. I don't think we will need a check for a specific docker version, as long as we have a good reason for it. 

I'm gonna test it a bit more in the afternoon, but it should work now as expect.

@shlomizadok & @ohadlevy mind giving it another test as well? :)

bastilian commented:
I moved away from the `Makefile` and added the `bin/foreman-container`, to maybe later have it in `$PATH` and allow starting development container from anywhere in the system mounting the current directory as foreman volume.

ohadlevy commented:
@bastilian anything i could help this move along? :)

bastilian commented:
Closing this as there is now a functioning and maintained docker environment for Foreman in the main repository.
